src/vmax.rs

```diff
@@ -1188,3 +1187,149 @@ impl ChildPoint for PhysicalCapacity {
     }
 }
+
+#[allow(non_snake_case)]
+#[derive(Debug, Deserialize, IntoPoint)]
+pub struct Volume {
+    #[serde(rename = "volumeId")]
+    pub volume_id: String,
+    #[serde(rename = "type")]
+    pub volume_type: Option<String>,
+    pub emulation: Option<String>,
+    pub ssid: Option<String>,
+    pub allocated_percent: Option<u64>,
+    pub cap_gb: Option<f64>,
+    pub cap_mb: Option<f64>,
+    pub cap_cyl: Option<u64>,
+    pub status: Option<String>,
+    pub reserved: Option<bool>,
+    pub pinned: Option<bool>,
+    pub physical_name: Option<String>,
+    pub volume_identifier: Option<String>,
+    pub wwn: Option<String>,
+    pub encapsulated: Option<bool>,
+    pub num_of_storage_groups: Option<u64>,
+    pub num_of_fron_end_paths: Option<u64>,
+    #[serde(rename = "storageGroupId")]
+    pub storage_group_id: Option<Vec<String>>,
+    // ignoring these two fields as they are not needed,
+    // and don't have to store an array unnecessarily
+    //#[serde(rename = "symmetrixPortKey")]
+    //pub symmetrix_port_key: Option<Vec<SymmetrixPortKey>>,
+    //#[serde(rename = "rdfGroupId")]
+    //pub rdf_group_id: Option<Vec<u64>>,
+    pub snapvx_source: Option<bool>,
+    pub snapvx_target: Option<bool>,
+    pub cu_image_base_address: Option<String>,
+    pub has_effective_wwn: Option<bool>,
+    pub effective_wwn: Option<String>,
+    pub encapsulated_wwn: Option<String>,
+}
+
+#[test]
+fn test_get_vmax_json_volume() {
+    use std::fs::File;
+    use std::io::Read;
+
+    let mut f = File::open("tests/vmax/slo_volume.json").unwrap();
+    let mut buff = String::new();
+    f.read_to_string(&mut buff).unwrap();
+
+    let i: Volume = serde_json::from_str(&buff).unwrap();
+    println!("result: {:#?}", i);
+    println!("point: {:#?}", i.into_point(Some("vmax_slo_volume")));
+}
+
+/// Returns a list of volume IDs for this array
+// TODO: combine this with the other getters and generalize along with other metrics
+pub fn get_all_slo_volumes(
+    client: &reqwest::Client,
+    config: &VmaxConfig,
+    symmetrixid: &str,
+) -> MetricsResult<Vec<String>> {
+    let data: Value = client
+        .get(&format!(
+            "https://{}/univmax/restapi/90/sloprovisioning/symmetrix/{}/volume",
+            config.endpoint, symmetrixid
+        ))
+        .basic_auth(&config.user, Some(&config.password))
+        .send()?
+        .error_for_status()?
+        .json()?;
+
+    let vol_count = match data["count"].as_u64() {
+        Some(count) => count,
+        None => 0,
+    };
+    let iterator_id = match data["id"].as_str() {
+        Some(id) => id,
+        None => "",
+    };
+    let max_count_per_page = match data["max_page_size"].as_u64() {
+        Some(count) => count,
+        None => 0,
+    };
+
+    if vol_count == 0 || max_count_per_page == 0 || iterator_id.is_empty() {
+        return Ok(vec![]);
+    }
+    let mut num_iterations = vol_count / max_count_per_page;
```

Comment:
I was surprised to learn this but this could actually be made more safe by doing `vol_count.checked_div(max_count_per_page)`.  It'll return None if there's a divide by zero instead of crashing the program.  https://play.rust-lang.org/?version=stable&mode=release&edition=2018&gist=a27e045f8d58980a14b5a31b2909ab4c.  So that aside I'm also wondering if this will miss the last page if it doesn't divide evenly.  Maybe it should be something like:
```
let num_pages = vol_count.checked_div(max_count_per_page).unwrap_or_else(|| 0);
let leftover = vol_count % max_count_per_page;
let mut num_iterations = num_pages + leftover;
```
maybe?

Reply:
The leftover is taken care because I start the loop after I have already retrieved the first page.

Reply:
Cool.  Sounds good